### PR TITLE
[remarkPluginAdjustLinkUrls] Pre-adjust folder-relative URLs

### PR DESCRIPTION
**Pull request by ashleemboyer**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-ambfolder-relative-urls-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1202114367927919/1203641801270773/f) 🎟️

## 🗒️ What

- Adds new `expandUrlTestCasesWithParams`
- Updates `src/lib/remark-plugins/rewrite-tutorial-links/__tests__/rewrite-external-docs-link.test.ts` to use the new helper
- Updates `preAdjustUrl` to handle folder-relative URLs
- Adds more test cases for `preAdjustUrl`, leveraging the new helper

## 🤷 Why

- Folder-relative URLs are currently broken on devdot, and they used to work on .io sites
- MDX content is being updated to no longer allow folder-relative URLs, but past versions of content will still have these URLs
- The new helper is useful for expanding URL-based test cases in multiple test suites, and having it as a function allows us to be more expressive about its use in a block comment + function name

### Current prod bug

- [ ] Go to `/vagrant/docs/plugins/go-plugins/guests#writing-a-guest-plugin` on [staging](https://dev-portal-git-staging-hashicorp.vercel.app/vagrant/docs/plugins/go-plugins/guests#writing-a-guest-plugin)
- [ ] Scroll just above the heading to the “Ruby plugins” link
- [ ] Inspect the URL
- [ ] Observe that the URL points to `/vagrant/guest-capabilities`
- [ ] Click the link
- [ ] Observe that you’re shown the “We couldn’t find the page you’re looking for” view

### This PR's expected behavior

- [ ] Go to `/vagrant/docs/plugins/go-plugins/guests#writing-a-guest-plugin` on [the preview](https://dev-portal-git-ambfolder-relative-urls-hashicorp.vercel.app/vagrant/docs/plugins/go-plugins/guests#writing-a-guest-plugin)
- [ ] Scroll just above the heading to the “Ruby plugins” link
- [ ] Inspect the URL
- [ ] Observe that the URL points to `/vagrant/docs/plugins/guest-capabilities`
- [ ] Click the link
- [ ] You should be taken to the correct docs page

### CODE REVIEW

File: src/lib/__tests__/remark-plugin-adjust-link-urls.test.ts [hunk 1]
```diff
@@ -2,21 +2,71 @@ import { preAdjustUrl } from 'lib/remark-plugin-adjust-link-urls'
...
+				'#heading',
+				'?param=value#heading',
 			]
-
 			test.each(urlsToTest)('"%s" is not pre-adjusted', (url: string) => {
 				expect(preAdjustUrl({ currentPath: mockCurrentPath, url })).toEqual(url)
 			})
 		})
 
+		describe('pre-adjusts folder-relative urls starting with `../`', () => {
```
Comment (kendallstrautman) [inline src/lib/__tests__/remark-plugin-adjust-link-urls.test.ts:19]: ❓just curious: have you found many folder relative urls like this in the content so far?

Comment (ashleemboyer) [reply src/lib/__tests__/remark-plugin-adjust-link-urls.test.ts:19]: Yep! Here's a Sourcegraph search that shows some: https://hashicorp-mktg.sourcegraph.com/search?q=context:global+lang:markdown+file:%5C.mdx%24+%5D%28../&patternType=standard&sm=0

File: src/lib/__tests__/remark-plugin-adjust-link-urls.test.ts [hunk 2]
```diff
@@ -1,46 +1,98 @@
+import {
+	expandUrlTestCasesWithParams,
+	TestCase,
+} from 'lib/testing/expand-url-test-cases-with-params'
 import { preAdjustUrl } from 'lib/remark-plugin-adjust-link-urls'
 
+const testEachCase = (testCases: TestCase[]) => {
+	const allCases = expandUrlTestCasesWithParams(testCases)
```
Comment (kendallstrautman) [inline src/lib/__tests__/remark-plugin-adjust-link-urls.test.ts:8]: 👍🏻

File: src/lib/remark-plugin-adjust-link-urls.ts [hunk 3]
```diff
@@ -36,6 +37,20 @@ export const preAdjustUrl = ({ currentPath, url }): string => {
 	const urlParts = url.split('/')
 	const currentPathParts = currentPath.split('/')
 
+	// Handle folder-relative URL that is linking upwards
+	if (url.startsWith('../')) {
+		const levelsUpCount = urlParts.filter((part) => part === '..').length
+		return path.join(
+			currentPathParts.slice(0, -(levelsUpCount + 1)).join('/'),
+			urlParts.slice(levelsUpCount).join('/')
```
Comment (kendallstrautman) [inline src/lib/remark-plugin-adjust-link-urls.ts:45]: Would be good to annotate with a note or extract into variables here to describe what these are doing.

File: src/lib/testing/expand-url-test-cases-with-params.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * Given a list of test cases, returns a new array with three additional case
+ * for each original case:
+ *
+ *  - the original URL + a query parameter string
+ *  - the original URL + an anchor hash
+ *  - the original URL + a query parameter string + an anchor hash
+ *
+ * @NOTE Any additional properties on each test case object WILL be preserved.
+ */
+const expandUrlTestCasesWithParams = (testCases: TestCase[]) => {
```
Comment (kendallstrautman) [inline src/lib/testing/expand-url-test-cases-with-params.ts:18]: This is great, ty!

**Review by kendallstrautman — APPROVED** [review]
Just a few comments on the path manipulation that could use some clarification. Non blocking though! Good work 🕵🏻 🚢 🎉

File: src/lib/remark-plugin-adjust-link-urls.ts (hunk 3 above)
Comment (ashleemboyer) [reply src/lib/remark-plugin-adjust-link-urls.ts]: Good idea! Just added two helpers with lots of comments through the steps.

- [`handleDotDotFolderRelativeUrl `](https://github.com/hashicorp/dev-portal/blob/amb.folder-relative-urls/src/lib/remark-plugin-adjust-link-urls.ts#L15-L45)
- [`handleDotSlashFolderRelativeUrl `](https://github.com/hashicorp/dev-portal/blob/amb.folder-relative-urls/src/lib/remark-plugin-adjust-link-urls.ts#L47-L67)

**Pull request merged**